### PR TITLE
[IMP] mail: redesign the chat views

### PR DESCRIPTION
**Pull request by Brieuc-brd**
This PR redesigns the content in the ChatWindow, Discuss and
the Chatter. The purpose is to make the chats more user-friendly.

Requires: 
- https://github.com/odoo/odoo/pull/95417

task-2823627

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Brieuc-brd** [comment]
Hello @alexkuhn @seb-odoo 👋 

This PR is finally ready for review 🙂 
Could you take a look at it?

cc. @stefanorigano 👀

File: addons/mail/static/src/components/message_in_reply_to_view/message_in_reply_to_view.xml [hunk 1]
```diff
@@ -2,22 +2,27 @@
...
+                            <b class="o_MessageInReplyToView_author">@<t t-out="messageInReplyToView.messageView.message.parentMessage.authorName"/></b>
+                            <span class="o_MessageInReplyToView_body text-break">
+                                <t t-if="messageInReplyToView.hasBodyBackLink">
+                                    <t t-out="messageInReplyToView.messageView.message.parentMessage.prettyBodyAsMarkup"/>
+                                </t>
+                                <t t-if="messageInReplyToView.hasAttachmentBackLink">
+                                    <span class="me-2 fst-italic">Click to see the attachments</span>
+                                    <i class="fa fa-image"/>
+                                </t>
+                            </span>
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message_in_reply_to_view/message_in_reply_to_view.xml:20]: As said in a prior comment: make a dedicated commit just for change of style for reply

File: addons/mail/static/src/components/message/message.xml [hunk 2]
```diff
@@ -53,20 +52,22 @@
...
                                 <t t-if="messageView.message.author">
-                                    <strong class="o_Message_authorName o_redirect me-2 text-truncate" t-att-class="{'o_cursor_pointer': messageView.hasAuthorOpenChat}" t-att-role="messageView.hasAuthorOpenChat ? 'button' : ''" t-att-tabindex="messageView.hasAuthorOpenChat ? '0' : ''" t-on-click="messageView.onClickAuthorName" t-att-title="messageView.authorTitleText">
+                                    <strong class="o_Message_authorName o_redirect me-1 text-truncate" t-att-class="{ 'o_cursor_pointer': messageView.hasAuthorOpenChat, 'd-none': messageView.inChatWindow and (messageView.message.isCurrentUserOrGuestAuthor or (messageView.messageListViewMessageViewItemOwner and  messageView.messageListViewMessageViewItemOwner.messageListViewOwner.thread.channel_type === 'chat')) }" t-att-role="messageView.hasAuthorOpenChat ? 'button' : ''" t-att-tabindex="messageView.hasAuthorOpenChat ? '0' : ''" t-on-click="messageView.onClickAuthorName" t-att-title="messageView.authorTitleText">
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:70]: Replace `d-none` by `t-if`

Comment (alexkuhn) [reply addons/mail/static/src/components/message/message.xml:70]: Also the condition is long and hard to understand at a glance.
Easily fixed by introducing the proper concept, e.g. with new (computed) field `MessageView/isShowingAuthorName`

File: addons/mail/static/src/components/message/message.xml [hunk 3]
```diff
@@ -111,54 +112,111 @@
...
-                        <div class="o_Message_content mx-2 text-break" t-ref="content">
-                            <t t-if="!messageView.composerViewInEditing">
-                                <div class="o_Message_prettyBody" t-ref="prettyBody"/><!-- messageView.message.prettyBody is inserted here from _update() -->
-                            </t>
-                            <t t-if="messageView.composerViewInEditing">
-                                <Composer
-                                    className="'o_Message_composer mb-1'"
-                                    record="messageView.composerViewInEditing"
+                        <div 
+                            class="o_Message_bubble_wrap position-relative d-flex align-items-start"
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:136]: rename `bubble_wrap` => `bubbleWrap` (camelCase)

File: addons/mail/static/src/components/message/message.xml [hunk 4]
```diff
@@ -111,54 +112,111 @@
...
+                            t-att-class="{
+                                'd-flex justify-content-end': messageView.inChatWindow and messageView.message.isCurrentUserOrGuestAuthor,
+                                'pe-5': messageView.inChatWindow and !messageView.message.isCurrentUserOrGuestAuthor and !messageView.composerViewInEditing,
+                                'ps-5': messageView.inChatWindow and messageView.message.isCurrentUserOrGuestAuthor and !messageView.composerViewInEditing,
+                                'pe-2': messageView.inChatWindow and !messageView.message.isCurrentUserOrGuestAuthor and messageView.composerViewInEditing,
+                                'ps-2': messageView.inChatWindow and messageView.message.isCurrentUserOrGuestAuthor and messageView.composerViewInEditing,
+                                'o_Message_can_be_deleted': messageView.message.canBeDeleted,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:143]: `o_Message_can_be_deleted` looks unused

File: addons/mail/static/src/components/message/message.scss [hunk 5]
```diff
@@ -52,21 +50,25 @@
...
-        border-bottom: 1px solid darken(map-get($grays, '400'), 5%);
+    &.o-reply-highlighted {
+        transform: translateY(-#{map-get($spacers, 3)});
     }
 
-    &.o-highlighted {
-        background-color: rgba($o-brand-primary, .1);
+    .o-not-current-author {
+        opacity: .15;
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.scss:61]: We did not follow our naming convention for these modifier classnames.
Since you are introducing and changing them, would camelCase e.g. `o-isReplyHighlighted` and `o-isAuthorNotCurrentUserOrGuest`

File: addons/mail/static/src/components/message/message.xml [hunk 6]
```diff
@@ -3,46 +3,45 @@
...
                 t-ref="root"
             >
-                <MessageInReplyToView t-if="messageView.messageInReplyToView" record="messageView.messageInReplyToView"/>
-                <div t-if="messageView.isActive and messageView.messageActionList" class="o_Message_actionListContainer position-absolute top-0 end-0 mt-n3 px-3" t-att-class="{ 'o-squashed': messageView.isSquashed }">
-                    <MessageActionList record="messageView.messageActionList"/>
+                <div class="o_Message_InReplyTo_wrap" t-att-class="{ 'me-5': messageView.inDiscuss or (messageView.inChatWindow and !messageView.message.isCurrentUserOrGuestAuthor), 'ms-5': messageView.inChatWindow and messageView.message.isCurrentUserOrGuestAuthor, 'd-flex justify-content-end': messageView.inChatWindow and messageView.message.isCurrentUserOrGuestAuthor }">
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:30]: rename `o_Message_InReplyTo_wrap` to `o_Message_inReplyToWrap`

File: addons/mail/static/src/components/message/message.xml (hunk 4 above)
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:143]: The template would be easier to understand conditional styling if you introduce concept of left/right alignment in models, e.g. `MessageView/isAlignedLeft` or `MessageView/isAlignedRight`.

**This comment applies to whole commit**

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/message_in_reply_to_view/message_in_reply_to_view.scss [hunk 7]
```diff
@@ -1,17 +1,27 @@
...
+.o_MessageInReplyToView_corner {
+    &.o-align-left {
+        left: $o-mail-message-sidebar-width / 2;
+        border-radius: $o-mail-rounded-rectangle-border-radius-sm 0 0 0;
+    }
+    
+    &.o-align-right {
+        right: $o-mail-message-sidebar-width / 2;
+        border-radius: 0 $o-mail-rounded-rectangle-border-radius-sm 0 0;
+    }
```
[the remark this answers is not in the document]
Comment (Brieuc-brd) [reply addons/mail/static/src/components/message_in_reply_to_view/message_in_reply_to_view.scss]: I already added the necessary models in the `MessageView` (https://github.com/odoo/odoo/pull/92872#discussion_r931137410)

What do you think ?

**Comment by Brieuc-brd** [comment]
Hello @alexkuhn 👋 ,
Thanks for your feedback, I updated the branch 👍 

What do you think ?

File: addons/web/static/src/legacy/scss/bootstrap_overridden.scss [hunk 8]
```diff
@@ -261,3 +261,6 @@ $form-switch-checked-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3
 $badge-color: inherit !default;
 $badge-padding-y: 0.25em !default;
 $badge-padding-x: 0.4em !default;
+
+// Placeholder color 
+$input-placeholder-color: $gray-300 !default;
```
Comment (alexkuhn) [inline addons/web/static/src/legacy/scss/bootstrap_overridden.scss:266]: This affects all input placeholders, so to see with @stefanorigano whether this is ok

File: addons/mail/static/src/models/attachment_list.js [hunk 9]
```diff
@@ -71,6 +71,66 @@ registerModel({
...
+        /**
+         * @private
+         * @returns {boolean}
+         */
+        _computeIsChatter() {
+            return Boolean(
+                (this.messageViewOwner && this.messageViewOwner.isChatter) ||
+                (this.composerViewOwner && this.composerViewOwner.isChatter)
+            );
+        },
```
Comment (alexkuhn) [inline addons/mail/static/src/models/attachment_list.js:103]: Missing `in` in-between `is` and `Discuss/ChatWindow/Chatter`
An Attachment List is no a X, but is in a X

File: addons/mail/static/src/models/composer_view.js [hunk 10]
```diff
@@ -903,6 +903,36 @@ registerModel({
...
+        /**
+         * @private
+         * @returns {boolean}
+         */
+        _computeIsChatter() {
+            return Boolean(
+                (this.threadView && this.threadView.threadViewer.chatter) ||
+                (this.messageViewInEditing && this.messageViewInEditing.isChatter)
+            );
+        },
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view.js:935]: Same remark as before: missing `in` in-between `is` and `X`

File: addons/mail/static/src/models/message_view.js [hunk 11]
```diff
@@ -267,6 +267,82 @@ registerModel({
...
+        /**
+         * @private
+         * @returns {boolean}
+         */
+        _computeIsChatter() {
+            return Boolean(
+                this.messageListViewMessageViewItemOwner &&
+                this.messageListViewMessageViewItemOwner.messageListViewOwner.threadViewOwner.threadViewer.chatter
+            );
+        },
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_view.js:302]: Same remark: missing `in` keyword

File: addons/mail/static/src/components/message/message.scss [hunk 12]
```diff
@@ -58,15 +56,21 @@
...
 
-        &.o_Message_active {
-            background-color: rgba($o-brand-primary, .2);
-        }
+    .o_Message_prettyBody a.o_mail_redirect {
+        padding: map-get($spacers, 1);
+        border-radius: $border-radius;
+        background: rgba($primary, .1);
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.scss:66]: unflatten, i.e. move it away from `.o_Message` rule and make it a top-level rule starting with `o_Message_prettyBody` immediately

File: addons/mail/static/src/models/message_in_reply_to_view.js [hunk 13]
```diff
@@ -44,6 +44,10 @@ registerModel({
         _computeHasBodyBackLink() {
             return !this.messageView.message.parentMessage.isBodyEmpty;
         },
+        /**
+         * @private
+         * @returns {boolean}
+         */
     },
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_in_reply_to_view.js:51]: useless diff

**Review by alexkuhn — COMMENTED** [review]
Comment just from looking at the code

File: addons/mail/static/src/components/message/message.xml (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/message/message.xml]: Ok for me, field name is a bit long, but I much prefer very explicit name than ambiguous or missing information, so that's good enough!

**Comment by alexkuhn** [comment]
Small remarks on visual:
1. chat window message should stick to message
chat-window-message-action-list-should-stick-to-message
2. mention style should pop more visually
mention-not-visible-enough
3. message action list should pop more visually
message-action-list-not-visible-enough
4. message action list anchor should have minimal gap with scrollbars
message-action-list-too-close-to-scrollbar

**Comment by Brieuc-brd** [comment]
> Small remarks on visual:
> 
> 1. chat window message should stick to message
> 
> chat-window-message-action-list-should-stick-to-message
> 
> 2. mention style should pop more visually
> mention-not-visible-enough
> 
> 3. message action list should pop more visually
> message-action-list-not-visible-enough
> 
> 4. message action list anchor should have minimal gap with scrollbars
> message-action-list-too-close-to-scrollbar

Hello @alexkuhn 👋 ,
I updated the PR.

About the visual remarks :

1. As we don't have enough space in the `ChatWindow`, we have to place the `MessageActionList` in front of the message when it’s too big (absolute position). 
If we stick the menu to the message, it will be partially hidden when the message is too long.
demo

To get a correct result, we should probably detect the length of the message to know in which position the menu should be. Or maybe with a css grid ? I did some tests but didn't get the expected result.
Maybe this could be an improvement for later, what do you think?

2. Opacity increased from `.1` to `.2` ✅ 

3. I added a light border, but I don't think we should do more to increase the contrast. This menu should not be more visible than the message bubble.

4.Fixed ✅

**Comment by alexkuhn** [comment]
Ok for keeping float message action list in chat window, because managing proper stick and overflow is quite tricky, and I prefer postponing this part until we have more awareness of UI data in JS modelling

File: addons/mail/static/src/components/message/message.scss [hunk 14]
```diff
@@ -52,21 +50,29 @@
...
-        border-bottom: 1px solid darken(map-get($grays, '400'), 5%);
+    &.o-isReplyHighlighted {
+        transform: translateY(-#{map-get($spacers, 3)});
     }
 
-    &.o-highlighted {
-        background-color: rgba($o-brand-primary, .1);
+    .o-isAuthorNotCurrentUserOrGuest {
+        opacity: .15;
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.scss:61]: should be
```scss
.o_Message_background {
    &.o-isAuthorNotCurrentUserOrGuest {
    }
}
```

File: addons/mail/static/src/components/message/message.xml [hunk 15]
```diff
@@ -3,46 +3,44 @@
...
                 }" t-attf-class="{{ messageView.extraClass }} {{ className }}" role="group" t-att-aria-label="messageView.message.messageTypeText" t-on-click="messageView.onClick" t-on-mouseenter="messageView.onMouseenter" t-on-mouseleave="messageView.onMouseleave" t-att-data-id="messageView.message.id" t-att-data-message-local-id="messageView.message.localId"
                 t-ref="root"
             >
-                <MessageInReplyToView t-if="messageView.messageInReplyToView" record="messageView.messageInReplyToView"/>
-                <div t-if="messageView.isActive and messageView.messageActionList" class="o_Message_actionListContainer position-absolute top-0 end-0 mt-n3 px-3" t-att-class="{ 'o-squashed': messageView.isSquashed }">
-                    <MessageActionList record="messageView.messageActionList"/>
+                <div class="o_Message_inReplyToWrap" t-att-class="{ 'me-5': messageView.isInDiscuss or messageView.isInChatWindowAndIsAlignedLeft, 'ms-5': messageView.isInChatWindowAndIsAlignedRight, 'd-flex justify-content-end': messageView.isInChatWindowAndIsAlignedRight }">
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:29]: can group `ms-5` and `d-flex justify-content-end` together (they have exact same condition)

File: addons/mail/static/src/components/message/message.xml [hunk 16]
```diff
@@ -3,46 +3,44 @@
...
+                    'o-squashed pt-1': messageView.isSquashed,
                     'o-starred': messageView.message.isStarred,
-                    'mt-3': !messageView.isSquashed and messageView.messageListViewMessageViewItemOwner,
-                    'bg-white': messageView.message.isDiscussionOrNotification and !messageView.isSelected and !(messageView.message.isHighlighted or messageView.isHighlighted) and !messageView.isActive,
-                    'bg-100': (!messageView.message.isDiscussionOrNotification and !messageView.isActive and !messageView.isSelected) or (messageView.isActive and !messageView.isSelected and messageView.message.isDiscussionOrNotification and !(messageView.message.isHighlighted or messageView.isHighlighted)),
-                    'bg-200': !messageView.message.isDiscussionOrNotification and messageView.isActive and !messageView.isSelected,
-                    'bg-400': messageView.isSelected and !(messageView.message.isHighlighted or messageView.isHighlighted),
+                    'o-currentAuthor': messageView.message.isCurrentUserOrGuestAuthor,
+                    'mt-1': !messageView.isSquashed,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:21]: Can group `o-squashed pt-1` and `mt-1` together (same condition)

File: addons/mail/static/src/components/message/message.xml [hunk 17]
```diff
@@ -111,54 +111,110 @@
...
+                                    t-if="messageView.message.prettyBody !== '' || messageView.composerViewInEditing"
+                                    class="o_Message_background position-absolute start-0 top-0 w-100 h-100"
+                                    t-att-class="{
+                                        'rounded-end-3': !(messageView.isInChatWindowAndIsAlignedRight),
+                                        'rounded-start-3': messageView.isInChatWindowAndIsAlignedRight,      
+                                        'rounded-bottom-3 border': messageView.message.isDiscussionOrNotification,
+                                        'rounded-3 opacity-50': !messageView.message.isDiscussionOrNotification,
+                                        'o-isAuthorNotCurrentUserOrGuest border-info bg-info-light': !messageView.message.isCurrentUserOrGuestAuthor and !messageView.message.isHighlighted and messageView.message.isDiscussionOrNotification, 
+                                        'border-success bg-success-light opacity-25': messageView.message.isCurrentUserOrGuestAuthor and !messageView.message.isHighlighted and messageView.message.isDiscussionOrNotification,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message.xml:164]: Can group `o-isAuthorNotCurrentUserOrGuest` and `border-success bg-success-light opacity-25` together (same condition)

**Comment by alexkuhn** [comment]
Some visual remarks in chatter:
1. I expect all messages to have a bubble. Maybe have basic greyed bubble
all-message-should-have-bubble-including-sms
2. message action list should stick to message in chatter
message-action-list-in-chatter-should-stick-to-message

File: addons/mail/static/src/components/message/message.xml (hunk 16 above)
Comment (Brieuc-brd) [reply addons/mail/static/src/components/message/message.xml:21]: `mt-1` has the opposite condition (`messageView.isSquashed` / `!messageView.isSquashed`).
We should not group them :)

File: addons/mail/static/src/components/message/message.xml (hunk 17 above)
Comment (Brieuc-brd) [reply addons/mail/static/src/components/message/message.xml]: Same as the previous one, `border-success ...` has `messageView.message.isCurrentUserOrGuestAuthor` in its condition while `o-isAuthorNotCurrentUserOrGuest` has the opposite.

File: addons/mail/static/src/components/message/message.xml (hunk 16 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/message/message.xml:21]: oops, nevermind 🤡

**Comment by Brieuc-brd** [comment]
> Some visual remarks in chatter:
> 
> 1. I expect all messages to have a bubble. Maybe have basic greyed bubble
>    all-message-should-have-bubble-including-sms
> 2. message action list should stick to message in chatter
>    message-action-list-in-chatter-should-stick-to-message

1. To keep the consistency, I kept the same colors for the sms 👍 .
Capture d’écran 2022-07-28 à 14 22 09

2. Same issue as for the ChatWindow (https://github.com/odoo/odoo/pull/92872#issuecomment-1197853554).
Keeping it sticky would create space and alignment issues (the number of icons is not always the same for all messages and could create an alignment inconsistency ) for large messages.
demo_2

File: addons/mail/static/src/components/message/message.xml (hunk 17 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/message/message.xml]: Indeed... ignore comment then

I have a hard time to see difference when there's `!` prefix on long expressions... I quite dislike the use of this transitive way of expression opposite of `x.y.z`, why language is not just `x.y.!z`😢

**Comment by alexkuhn** [comment]
Log note missing bubble:
Screenshot 2022-07-28 at 14 35 39

I think would make sense to have grey bubble for this one.
SMS is okay to keep same style as conversation, because there's the SMS icon on top of message.
Log note, however, there was just background being greyed, so make sense to have different bubble style?

**Comment by Brieuc-brd** [comment]
> Log note missing bubble: Screenshot 2022-07-28 at 14 35 39
> 
> I think would make sense to have grey bubble for this one. SMS is okay to keep same style as conversation, because there's the SMS icon on top of message. Log note, however, there was just background being greyed, so make sense to have different bubble style?

After internal discussion, this choice was made to avoid adding another bubble color and to keep the rendering less "noisy" as possible.
Moreover, it makes the messages more visible in the thread.

**Comment by alexkuhn** [comment]
> After internal discussion, this choice was made to avoid adding another bubble color and to keep the rendering less "noisy" as possible. Moreover, it makes the messages more visible in the thread.

IMO this is not a good choice.
When you have bubble layout, I expect all exchange of messages (notes is an internal exchange of messages) to all have this bubble visual.

So I pretty much disagree with this decision 🫤

**Comment by Brieuc-brd** [comment]
@alexkuhn,
I just fixed a small condition issue due to the last push and added a `text-muted` on log notes to increase the contrast with a regular message.

**Comment by alexkuhn** [comment]
> If I'm talking with a client I prefer to see my notes and activities designed very differently from the messages I'm exchanging with him.

Agree, it's important to distinguish both, because the context is quite different.
TBH I think it's hard to make it different while having a consistent shape e.g. with also bubble layout.
So maybe it's the best trade-off to not have bubble visual for notes.

It's a shame nonetheless, because most internal communication we do is with notes, so we won't have bubble for the most part. Also notes are quite noisy: never squashed, so they always show avatar + name for each of them.

I'm wondering whether the exchange of notes could be easier to see with a change of conversation background just for them... And also prevent the squash.

Here is an idea:
1. currently in this branch:
chat-window-current
2. suggestion:
chat-window-suggestion



Personally, I really like when I can understand the overall shape just at a glance. That's why having some visual clues like this would really improvement usability for me. Thanks!

**Comment by Brieuc-brd** [comment]
Hello @alexkuhn 👋 ,

I have updated the rendering of the log notes:
- Same padding as the bubble for consistency.
- Fully rounded corner, so it's not a "conversation bubble" but rather a "voiceover bubble".
- Light gray background: slightly visible but not too noisy.
- No border, let's leave the border for real messages.
- Full width: it allows to take out the notes of the conversation.
Capture d’écran 2022-07-29 à 09 03 13

I also improved the design of the TrackingValue in a separate commit.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+ rebase-ff

File: addons/web/static/src/legacy/scss/bootstrap_overridden.scss (hunk 8 above)
Comment (stefanorigano) [reply addons/web/static/src/legacy/scss/bootstrap_overridden.scss:268]: Thanks @alexkuhn .
@Brieuc-brd , I missed this notification. `$gray-300` is sensible lower than default (600)... have you tested the entire UI?

Comment (Brieuc-brd) [reply addons/web/static/src/legacy/scss/bootstrap_overridden.scss:268]: @stefanorigano Yes, this variable seems to be used only for inputs with the `form-control` class. 
https://github.com/odoo/odoo/blob/d108db36ca2974179ded929370ea45a106f2e08c/addons/web/static/lib/bootstrap/scss/forms/_form-control.scss#L57

The purpose of this change is to have the same placeholder color with or without this class to keep consistency.
eg. input with placeholder without the `form-control` class (Project app):
Capture d’écran 2022-07-29 à 14 45 01
